etc/generate_tiles.py

```diff
@@ -0,0 +1,207 @@
+
+        # Render image with default Agg renderer
+        im = mapnik.Image(render_size, render_size)
+        mapnik.render(self.m, im)
+        im.save(tile_uri, 'png256')
+
+
+    def loop(self):
+        while True:
+            #Fetch a tile from the queue and render it
+            r = self.q.get()
+            if (r == None):
+                self.q.task_done()
+                break
+            else:
+                (name, tile_uri, x, y, z) = r
+
+            exists= ""
+            if os.path.isfile(tile_uri):
+                exists= "exists"
+            else:
+                self.render_tile(tile_uri, x, y, z)
+            bytes=os.stat(tile_uri)[6]
+            empty= ''
+            if bytes == 103:
+                empty = " Empty Tile "
+            self.printLock.acquire()
+            print name, ":", z, x, y, exists, empty
+            self.printLock.release()
+            self.q.task_done()
+
+
+
+def render_tiles(bbox, mapfile, tile_dir, minZoom=1,maxZoom=18, name="unknown", num_threads=NUM_THREADS, tms_scheme=False):
+    print "render_tiles(",bbox, mapfile, tile_dir, minZoom,maxZoom, name,")"
+
+    # Launch rendering threads
+    queue = Queue(32)
+    printLock = threading.Lock()
+    renderers = {}
+    for i in range(num_threads):
+        renderer = RenderThread(tile_dir, mapfile, queue, printLock, maxZoom)
+        render_thread = threading.Thread(target=renderer.loop)
+        render_thread.start()
+        #print "Started render thread %s" % render_thread.getName()
+        renderers[i] = render_thread
+
+    if not os.path.isdir(tile_dir):
+         os.mkdir(tile_dir)
+
+    gprj = GoogleProjection(maxZoom+1) 
+
+    ll0 = (bbox[0],bbox[3])
+    ll1 = (bbox[2],bbox[1])
+
+    for z in range(minZoom,maxZoom + 1):
+        px0 = gprj.fromLLtoPixel(ll0,z)
+        px1 = gprj.fromLLtoPixel(ll1,z)
+
+        # check if we have directories in place
+        zoom = "%s" % z
+        if not os.path.isdir(tile_dir + zoom):
+            os.mkdir(tile_dir + zoom)
+        for x in range(int(px0[0]/256.0),int(px1[0]/256.0)+1):
+            # Validate x co-ordinate
+            if (x < 0) or (x >= 2**z):
+                continue
+            # check if we have directories in place
+            str_x = "%s" % x
+            if not os.path.isdir(tile_dir + zoom + '/' + str_x):
+                os.mkdir(tile_dir + zoom + '/' + str_x)
+            for y in range(int(px0[1]/256.0),int(px1[1]/256.0)+1):
+                # Validate x co-ordinate
+                if (y < 0) or (y >= 2**z):
+                    continue
+                # flip y to match OSGEO TMS spec
+                if tms_scheme:
+                    str_y = "%s" % ((2**z-1) - y)
+                else:
+                    str_y = "%s" % y
+                tile_uri = tile_dir + zoom + '/' + str_x + '/' + str_y + '.png'
+                # Submit tile to be rendered into the queue
+                t = (name, tile_uri, x, y, z)
```

Comment:
Can we split out the bit which generates all of the (tile URIs, x, y, z) tuples into a separate function? Then, this can be simplified:
```python
for (tile_uri, x, y, z) in tile_uris():
    ensure_directory(tile_uri)
    queue.put((name, tile_uri, x, y, z))
```

Reply:
In general, this function is pretty long and does a lot. It would be easier to read if it were split into smaller functions which do a single thing.